### ISSUE TITLE
QueryFilter is not being applied to related entities

### ISSUE THREAD

**Issue by sebastianroesch** [zzzprojects/EntityFramework-Plus#26]
QueryFilter is not being applied to related entities. When I define a soft-delete filter, it works only when accessing the main entity set of the context. But when I access child collections of an entity, the filter is not applied.

Our Query Filter is defined like this: 

`ctx.Filter<HQBaseEntity>("IsDeleted", q => q.Where(x => !x.DeletedOn.HasValue));`

It generally works, because the DeletedOn IS NULL is correctly applied to the WHERE clause when accessing entities like this:

`var nonDeletedDocument = ctx.Documents.Where(d => d.Id == 4101);`

```
SELECT TOP (2) 
    [Extent1].[Id] AS [Id], 
    [Extent1].[ShownDate] AS [ShownDate], 
    [Extent1].[ValidUntilDate] AS [ValidUntilDate], 
    [Extent1].[SentOn] AS [SentOn], 
....
    [Extent1].[UpdatedOn] AS [UpdatedOn], 
    [Extent1].[UpdatedBy] AS [UpdatedBy], 
    [Extent1].[DeletedOn] AS [DeletedOn], 
    [Extent1].[DeletedBy] AS [DeletedBy]
    FROM [dbo].[Document] AS [Extent1]
    WHERE ([Extent1].[DeletedOn] IS NULL) AND ([Extent1].[Id] = @p__linq__0)
-- p__linq__0: '4101' (Type = Int32, IsNullable = false)
```

However, EF-Plus Query Filter does not apply our Deleted filter when accessing entities like this:

`var nonDeletedPositionLinks = nonDeletedDocument.DocumentPositionToDocuments;`

Here is the query generated by EF-Plus Query Filter when accessing child collections. It is missing the WHERE clause for the DELETED IS NULL:

```
SELECT 
    [Extent1].[Id] AS [Id], 
    [Extent1].[Amount] AS [Amount], 
    [Extent1].[Order] AS [Order], 
    [Extent1].[IsCompleted] AS [IsCompleted], 
    [Extent1].[Text] AS [Text], 
....
    [Extent1].[UpdatedOn] AS [UpdatedOn], 
    [Extent1].[UpdatedBy] AS [UpdatedBy], 
    [Extent1].[DeletedOn] AS [DeletedOn], 
    [Extent1].[DeletedBy] AS [DeletedBy]
    FROM [dbo].[DocumentPositionToDocument] AS [Extent1]
    WHERE [Extent1].[DocumentId] = @EntityKeyValue1
-- EntityKeyValue1: '4101' (Type = Int32, IsNullable = false)
```

We were using DynamicFilters (https://github.com/jcachat/EntityFramework.DynamicFilters) before, which applies the query filter also to child collections. This filter fails, however, when used in combination with EF+.

**Comment by zzzprojects** [comment]
Hello Sebastian,

Just to let you know we have not abandoned this issue. We are still looking how to make it work.

Unfortunately, we have a difference strategy than EntityFramework.DynamicFilters, so we are still searching how to achieve it within our library.

**Comment by sebastianroesch** [comment]
Hi, thank you! If there is any way I can help, let me know.

**Comment by sebastianroesch** [comment]
Hi, are there any updates on this issue? We're about to purchase EF+ for the Batch SaveChanges, but are currently not able to completely use it due to this filter issue.

**Comment by zzzprojects** [comment]
Hello Sebastian,

I will try to pass a full day about the issue this weekend. I will give you an answer shortly.

**Comment by zzzprojects** [comment]
Hello Sebastian,

We are making great progress!

Unfortunately for us, Lazy Loading do not use LINQ query.

Here is basically how lazy loading work:
1. Generate a template SQL text using some entity field
2. Parse the template to an expression tree
3. Use DbExpressionVisitor interceptor
4. Generate the SQL from the expression tree

The only place we can change the expression tree is at step 3 as EntityFramework.DynamicFilters do. 

However, we don't want to use the same technic as EntityFramework.DynamicFilters do since they create themselves the filter for the DbExpression. So as soon as someone is using something they don't support, that break the code.

We instead try to use the filtered expression and let Entity Framework do all the job.

We successfully make it work with some basic LINQ method like Where, Take, etc. however the code is still at prototype step.

We will continue to work on it today. I'm pretty sure by the end of the day, I will be able to provide a release date or maybe a beta version.

**Comment by sebastianroesch** [comment]
Hi, 
that sounds great, can't wait to try it out! 
Thanks a lot!

**Comment by zzzprojects** [comment]
Hello Sebastian,

Yesterday we faced a very strange issue. I worked until 5AM to try out to understand it unsuccessfully. That's why we have not released a version yesterday.

However, today, we successfully fixed it and all the code has passed all our unit test.

We are currently cleaning up the code. I must leave in less than one hour, so I will try to release a version this night if I'm not back to late otherwise tomorrow morning.

**Comment by zzzprojects** [comment]
Hello Sebastian,

As you probably are already aware, the EF+ Query Filter has been completely re-written and work differently from the previous version.

We are proud to announce you a beta version is now available (v1.3.13-beta)
https://www.nuget.org/packages/Z.EntityFramework.Plus.EF6/1.3.13-beta

There is still some refactoring work to do and to make sure to better manage concurrency scenario but this version should be enough stable to let you test all the filter features.

It works exactly like the previous version, only the **QueryFilterManager.InitilizeGlobalFilter** method will probably be removed on the stable version since it's now obsolete but is still exists in the BETA version.

We will wait for your feedback before releasing a stable version. From our side, we will complete the refactoring this weekend since we are late in our roadmap for this week.

**Comment by sebastianroesch** [comment]
Hi, thanks for the effort! I tried it out today and it works for the related entities!
However, now I am receiving errors from what seems as within EF+:

"DbExpressionBinding requires an input expression with a collection ResultType.\r\nParameter name: input"
   at System.Data.Entity.Core.Common.CommandTrees.ExpressionBuilder.DbExpressionBuilder.BindAs(DbExpression input, String varName)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpressionBinding(DbExpressionBinding binding)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpressionBindingEnterScope(DbExpressionBinding binding)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbFilterExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DbFilterExpression.Accept[TResultType](DbExpressionVisitor`1 visitor)
   at Z.EntityFramework.Plus.DbScanExpressionVisitor.ApplyFilter(DbExpression baseExpression, String fullName)
   at Z.EntityFramework.Plus.DbScanExpressionVisitor.Visit(DbPropertyExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DbPropertyExpression.Accept[TResultType](DbExpressionVisitor`1 visitor)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitUnary(DbUnaryExpression expression, Func`2 callback)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbIsNullExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DbIsNullExpression.Accept[TResultType](DbExpressionVisitor`1 visitor)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitUnary(DbUnaryExpression expression, Func`2 callback)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbNotExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DbNotExpression.Accept[TResultType](DbExpressionVisitor`1 visitor)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbFilterExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DbFilterExpression.Accept[TResultType](DbExpressionVisitor`1 visitor)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.VisitExpression(DbExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DefaultExpressionVisitor.Visit(DbLimitExpression expression)
   at System.Data.Entity.Core.Common.CommandTrees.DbLimitExpression.Accept[TResultType](DbExpressionVisitor`1 visitor)
   at Z.EntityFramework.Plus.QueryInterceptorFilterDbCommandTree.TreeCreated(DbCommandTreeInterceptionContext interceptionContext)
   at System.Data.Entity.Infrastructure.Interception.DbCommandTreeDispatcher.<Created>b__0(IDbCommandTreeInterceptor i, DbCommandTreeInterceptionContext c)
   at System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch[TInterceptionContext,TResult](TResult result, TInterceptionContext interceptionContext, Action`2 intercept)
   at System.Data.Entity.Infrastructure.Interception.DbCommandTreeDispatcher.Created(DbCommandTree commandTree, DbInterceptionContext interceptionContext)
   at System.Data.Entity.Core.Common.DbProviderServices.CreateCommandDefinition(DbCommandTree commandTree, DbInterceptionContext interceptionContext)
   at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlanFactory.CreateCommandDefinition(ObjectContext context, DbQueryCommandTree tree)


The Linq query is similar to this: ctx.ContactPersons.FirstOrDefault(c => c.Company != null);

Let me know if you need any more information on this.

**Comment by zzzprojects** [comment]
This issue is fixed in the v1.3.14
https://www.nuget.org/packages/Z.EntityFramework.Plus.EF6/1.3.14-beta

However, that bring a whole new kind of filter!

We normally filter on a collection to let using Where, Take, Skip, etc. clause.

In this situation, we don't filter anymore a collection but directly the entity! Filtering directly an entity is still not supported but I will look at this issue this night.

I believe if we really want to support some situation fully, we should be able also to filter the entity directly when a LINQ like the one you provided is ussed.

**Comment by zzzprojects** [comment]
Closing: Issue has been fixed in v1.3.14
